### PR TITLE
Auth: implement auto_sign_up for auth.jwt

### LINKED ISSUE #34497: JWT Auth - Add Signing up users config as on many other auth modes
**What would you like to be added**:
Related to https://github.com/grafana/grafana/issues/8198
Permit to add the config field `auto_sign_up` like on any other auth mode for JWT.

### PR DESCRIPTION
**Pull request by Roguelazer**
Uses the RFC 7519-mandated "sub" field as the external Auth ID. If present, uses the IANA-registered "name" field as the full name of the auto-created user.

**What this PR does / why we need it**: Adds the `auto_sign_up` option to JWT authentication, matching other auth providers

**Which issue(s) this PR fixes**: Fixes #34497

### CODE REVIEW

File: docs/sources/auth/jwt.md
Comment (achatterjee-grafana) [inline docs/sources/auth/jwt.md:50]:
```suggestion
If `auto_sign_up` is enabled, the `sub` claim is used as the "external Auth ID" and the `name` claim, is present, is used as the user's full name.
```

**Review by achatterjee-grafana — COMMENTED** [review]
Added doc review, needs Dev review before approval.

**Comment by achatterjee-grafana** [comment]
@marefr - can you review when you have a moment. Thanks.

**Review by mwaldt — APPROVED** [review]

File: docs/sources/auth/jwt.md
Comment (achatterjee-grafana) [inline docs/sources/auth/jwt.md:50]:
```suggestion
If `auto_sign_up` is enabled, then the `sub` claim is used as the "external Auth ID". The `name` claim is used as the user's full name if it is present.
```

**Review by achatterjee-grafana — COMMENTED** [review]
Minor copy edit suggestion.

**Review by achatterjee-grafana — APPROVED** [review]

**Comment by Roguelazer** [comment]
Anything y'all would like me to do to move this forward?

**Comment by Roguelazer** [comment]
@pkolyvas it looks like this is blocked on your review?

**Comment by gillg** [comment]
Cool ! I missed this PR, I hope it will be merged soon !
In addition, make json queries to add a group mapping or choose the login attributes would be nice. Could be in a second time if this PR is merged soon.

**Review by sakjur — APPROVED** [review]

**Comment by sakjur** [comment]
Sorry for this review taking so much time, I was unaware of this PR. It looks really good, and I'd be happy to merge this for inclusion in 8.3. 🙈

@Roguelazer Could you merge `main` to this PR and push? I've taken a look at this and I'm happy to merge it in its current state.

One minor problem right now is that when you create these users they show up as "OAuth" in the admin users list, I don't think that's a blocker, but if you could apply this patch to your PR that would fix that:

```patch
commit 183a0eccec9cfb30884641d08c544fdb05a3aef7
Author: Emil Tullstedt <emil.tullstedt@grafana.com>
Date:   Thu Nov 4 13:01:50 2021 +0100

    Auth: Add label for JWT in users list

diff --git a/pkg/services/searchusers/searchusers.go b/pkg/services/searchusers/searchusers.go
index c7070d26ce..2bb1c4dbcb 100644
--- a/pkg/services/searchusers/searchusers.go
+++ b/pkg/services/searchusers/searchusers.go
@@ -96,6 +96,8 @@ func GetAuthProviderLabel(authModule string) string {
 		return "SAML"
 	case "ldap", "":
 		return "LDAP"
+	case "jwt":
+		return "JWT"
 	default:
 		return "OAuth"
 	}
```

**Comment by bizrad** [comment]
> @Roguelazer Could you merge `main` to this PR and push?

Is this still waiting on the merge, push and change mentioned in the last comment? Just curious as we are looking forward to getting this change in for our new setup.

**Comment by sakjur** [comment]
@bizrad Thanks for the nudge. Yes it is, I just opened #43502 as an alternative mergeable path.

I'd prefer to merge this PR, but I don't think I've been given push permissions so I can't merge main to it.

**Comment by Roguelazer** [comment]
I rebased this on main and applied the commit from @sakjur.

**Review by sakjur — APPROVED** [review]

**Comment by Roguelazer** [comment]
I can't say I quite understand how my diff caused that particular test to fail...

**Comment by sakjur** [comment]
Agreed, I’ll try to figure out what’s up with that test.

**Comment by sakjur** [comment]
Merged for 8.4.0 through #43502. 🙏 Thanks for the PR!

**Pull request closed without merging**
